MapboxDirections/OfflineDirections.swift

```diff
@@ -0,0 +1,139 @@
+public typealias OfflineVersionsHandler = (_ version: [Version]?, _ error: Error?) -> Void
+
+struct AvailableVersionsResponse: Codable {
+    let availableVersions: [Version]
+}
+
+@objc(MBVersion)
+public class Version: NSObject, Codable {
+    let versionString: String
+    let versionDate: Date
+    
+    static var dateFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "yyyy-MM-dd"
+        return formatter
+    }()
+    
+    public convenience init(_ versionDate: Date) {
+        self.init(versionString: Version.dateFormatter.string(from: versionDate), versionDate: versionDate)
+    }
+    
+    public convenience init(_ versionString: String) {
+        self.init(versionString: versionString, versionDate: Version.dateFormatter.date(from: versionString)!)
+    }
+    
+    init(versionString: String, versionDate: Date) {
+        self.versionString = versionString
+        self.versionDate = versionDate
+    }
+    
+    required public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let versionString = try container.decode(String.self)
+        
+        self.versionString = versionString
+        self.versionDate = Version.dateFormatter.date(from: versionString)!
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+        try container.encode(versionString)
+    }
+}
+
+@objc(MBOfflineDirectionsProtocol)
+public protocol OfflineDirectionsProtocol {
+    
+    /**
+     Fetches the available versions.
+     */
+    @discardableResult
+    func availableOfflineVersions(completionHandler: @escaping OfflineVersionsHandler) -> URLSessionDataTask
+    
+    /**
+     Initiates a download process of all tiles needed to provide routing within the given bounding box.
+     
+     - parameter boundingBox: The region of the pack to be downloaded.
+     - parameter version: The version of the pack to be downloaded.
+     - parameter progressHandler: Reports the progress of downloaded and yet to be downloaded bytes
+     - parameter completionHandler: Informs when the download is completed or failed. The offline pack may be moved from the temporary directory and to a persistent store at this point.
+     */
+    @discardableResult
+    func downloadTiles(for boundingBox: BoundingBox, version: Version, progressHandler: @escaping OfflineDownloaderProgressHandler, completionHandler: @escaping OfflineDownloaderCompletionHandler) -> URLSessionDownloadTask
+}
+
+extension Directions: OfflineDirectionsProtocol, URLSessionDownloadDelegate {
+    
+    func availableVersionsURL() -> URL {
+        
+        let url = apiEndpoint.appendingPathComponent("route-tiles/v1").appendingPathComponent("versions")
+        var components = URLComponents(string: url.absoluteString)
```

Comment:
Nit: use `URLComponents(url:resolvingAgainstBaseURL:)` for better type safety. (×2)